File: Lista/quest09/interfacesprotocolos.go
package main

import "fmt"

type Imprimivel interface {
	Imprimir()
}

type Relatorio struct {
	Titulo string
}

func (r Relatorio) Imprimir() {
	fmt.Println("Imprimindo Relatório:", r.Titulo)
}

type Contrato struct {
	ParteA string
	ParteB string
}

func (c Contrato) Imprimir() {
	fmt.Printf("Imprimindo Contrato entre %s e %s\n", c.ParteA, c.ParteB)
}

func main() {
	var documentos []Imprimivel

	documentos = append(documentos, Relatorio{Titulo: "Relatório Anual"})
	documentos = append(documentos, Contrato{ParteA: "Empresa A", ParteB: "Empresa B"})

	for _, doc := range documentos {
		doc.Imprimir()
	}
}

File: Lista/quest09/interfacesprotocolos.py
from abc import ABC, abstractmethod

class Imprimivel(ABC):
    @abstractmethod
    def imprimir(self):
        pass

class Relatorio(Imprimivel):
    def __init__(self, titulo):
        self.titulo = titulo

    def imprimir(self):
        print(f"Imprimindo Relatório: {self.titulo}")

class Contrato(Imprimivel):
    def __init__(self, parte_a, parte_b):
        self.parte_a = parte_a
        self.parte_b = parte_b

    def imprimir(self):
        print(f"Imprimindo Contrato entre {self.parte_a} e {self.parte_b}")

if __name__ == "__main__":
    documentos = [
        Relatorio("Relatório Anual"),
        Contrato("Empresa A", "Empresa B")
    ]

    for doc in documentos:
        doc.imprimir()
